File: BinaryTree/univalued_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isUnivalTree(self, root: Optional[TreeNode]) -> bool:
        vals = set()
        def helper(root, vals):
            if root:
                vals.add(root.val)
                helper(root.left, vals)
                helper(root.right, vals)
        helper(root, vals)
        return len(vals) == 1


    # shorter
    def isUnivalTree(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            return not node or node.val == root.val and dfs(node.left) and dfs(node.right)
        return dfs(root)

File: BinaryTree/CPP/univalued_binary_tree.cpp
#include <set>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isUnivalTree(TreeNode* root) {
        set<int> values;
        helper(root, values);
        return values.size() == 1;
    }

    void helper(TreeNode* root, set<int>& values) {
        if (root) {
            values.insert(root->val);
            helper(root->right, values);
            helper(root->left, values);
        }
    }

    // shorter
    bool isUnivalTree(TreeNode* root, int val = -1) {
        if (!root) return true;
        if (val < 0) val = root-> val;
        return root->val == val && isUnivalTree(root->left, val) && isUnivalTree(root->right, val);
    }

    // one-liner
    // bool isUnivalTree(TreeNode* root) {
    //     return (!root->left || root->left->val == root->val && isUnivalTree(root->left)) &&
    //            (!root->right || root->right->val == root->val && isUnivalTree(root->right));
    // }
};